apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,921 @@
+ *   #/composer?recipient=+33615647897 => composer
+ *   #composer                          => composer
+ *
+ * @param {String} hash The string to analyze.
+ * @returns {String} The analyzed string. It is not guaranteed to be an existing
+ * view name.
+ */
+function findViewNameFromHash(hash) {
+  debug('findViewNameFromHash(%s)', hash);
+  if (!hash) {
+    return null;
+  }
+
+  if (hash[0] === '#') {
+    hash = hash.slice(1);
+  }
+
+  var index = hash.indexOf('/');
+  if (index !== -1) {
+    hash = hash.slice(index + 1);
+  }
+
+  index = hash.indexOf('?');
+  if (index !== -1) {
+    hash = hash.slice(0, index);
+  }
+
+  return hash;
+}
+
+/**
+ * Tries to find the correct view from the location.
+ *
+ * First it tries to find the view using the location's hash.
+ * If findViewNameFromHash returns something, we try to match an existing view
+ * from this name, then we check whether the found view is indeed part of this
+ * document using the full URL.
+ *
+ * If findViewNameFromHash returns nothing, we try to find a view from the full
+ * URL.
+ *
+ * @param {Location} location The content of window.location.
+ * @returns {Object} The found view, or null.
+ */
+function findViewFromLocation(location) {
+  debug('findViewFromLocation(%s)', location.href);
+
+  var pathName = location.pathname;
+  if (pathName.endsWith('/')) {
+    pathName += 'index.html';
+  }
+
+  var hash = location.hash;
+  var viewNameFromHash = findViewNameFromHash(hash);
+  var viewFromHash = findViewFromName(viewNameFromHash);
+  if (viewNameFromHash && !viewFromHash) {
+    console.error('We found no view from the hash %s', hash);
+    return null;
+  }
+
+  var viewFromPath = VIEWS.find(
+    (view) => pathName.endsWith('/' + view.url)
+  ) || null;
+
+  if (!viewFromPath) {
+    return null;
+  }
+
+  if (viewFromHash && viewFromHash.partOf &&
+      viewFromHash.partOf !== viewFromPath.name) {
     console.error(
-      'Navigation.toPanel: got an exception: (%s) %s',
-      e, e.stack && e.stack.replace(/\n/g, '|')
+      'The view %s found from the hash %s is not part of view %s',
+      viewFromHash.name,
+      location.hash,
+      viewFromPath.name
     );
-  } else {
-    console.log('Navigation.toPanel: got a rejected promise');
+    viewFromHash = null;
   }
+
+  return viewFromHash || viewFromPath;
+}
+
+/**
+ * Finds a view from its name.
+ *
+ * @param {String} name Name of the view to find.
+ * @returns {Object} The found view, or null.
+ */
+function findViewFromName(name) {
+  debug('findViewFromName(%s)', name);
+  var view = VIEWS.find((view) => view.name === name) || null;
+
+  debug('findViewFromName() =>', view && view.name);
+  return view;
 }
 
-var currentPanel;
-var queuedPanel;
+/**
+ * Finds the main view of a view.
+ *
+ * @param {Object} view
+ * @returns {Object} the main view of the parameter view. It can be the same
+ * object if the parameter is itself the main view.
+ */
+function findMainView(view) {
```

Comment:
>  But I'd like to use another term than "master" as it can be considered as offensive by some people.

Oh, crazy world :)

What about "host/container"? If it doesn't work, let's move on with main then